[VPD-479]: Swap Router

Debugger022 opened the pull request:
This PR introduces the `SwapRouter` contract to the Venus periphery, along with its interface, deployment script, and supporting tests. The contract is now deployed and verified on both `bsctestnet` and `bscmainnet`.

What’s Included

* `SwapRouter` Contract
    * Implements swap-and-supply and swap-and-repay operations for Venus markets.
    * Supports both ERC-20 and native token (BNB) flows.
    * Includes robust error handling, slippage protection, and event emission for all major actions.
    * Uses upgradeable and secure patterns (Ownable2StepUpgradeable, ReentrancyGuardUpgradeable).
    
* `ISwapRouter` Interface
    * Defines the external interface for all SwapRouter operations, including supply, repay, and sweep functions.
    
* `Deployment Script`
    * [003-deploy-swap-router.ts] automates deployment with proxy support and ownership transfer to the timelock.
    * Handles all required dependencies and admin setup for mainnet and testnet.
    
* `Tests`
    * Comprehensive tests for SwapRouter functionality, covering all major flows and error cases.
    
* `Network Deployments`
    * SwapRouter is deployed and verified on both `bsctestnet` and `bscmainnet`

kkirka committed:
feat: add generic swapper contract

kkirka committed:
feat: add backend signature verification to the swapper

Debugger022 committed:
feat: add SwapRouter contract

Debugger022 committed:
test: add unit test for SwapRouter

Debugger022 committed:
feat: add deployment script

Debugger022 committed:
feat: add SwapRouter fork test

GitGuru7 commented on contracts/SwapRouter/SwapRouter.sol:
```diff
@@ -0,0 +1,493 @@
+    IWBNB public immutable wrappedNative;
+
+    /// @notice The native vToken address (e.g., vBNB)
+    address public immutable NATIVE_VTOKEN;
```
For consistency with other immutable contract references, can we use the `IVBNB` interface here as well?

GitGuru7 commented on contracts/SwapRouter/SwapRouter.sol:
```diff
@@ -0,0 +1,493 @@
+     */
+    function _getUnderlyingToken(address vToken) internal view returns (address underlying) {
+        if (vToken == NATIVE_VTOKEN) {
+            return address(wrappedNative);
```
I understand we’re returning the wrapped asset here instead of the native one since the native token doesn’t have an address and it simplifies the swap logic. This is later handled in `_supply()`.
However, should we continue supporting this mapping for the native market? Since we plan to deprecate the native market and move fully to the wrapped version, should we all together  skip the vBNB minting wdyt?

Debugger022 replied:
I agree with this approach, and it's worth discussing further.

GitGuru7 commented on contracts/SwapRouter/SwapRouter.sol:
```diff
@@ -0,0 +1,493 @@
+        uint256 amountIn,
+        uint256 minAmountOut,
+        bytes[] calldata swapData
+    ) external payable nonReentrant {
```
Any specific reason for making `swapAndSupply` and `swapAndRepay` payable when native handling already uses separate functions (`swapNativeAndSupply` and `swapNativeAndRepay`)?

GitGuru7 commented on contracts/SwapRouter/SwapRouter.sol:
```diff
@@ -0,0 +1,493 @@
+        if (amountOut > debtAmount) {
+            uint256 excess = amountOut - debtAmount;
+            IERC20Upgradeable(tokenOut).safeTransfer(msg.sender, excess);
+        }
```
Nit: Can we refactor the logic so excess tokens are transferred back to the user in one place instead of doing it twice, once here and once inside `_repay`?
I understand the calculations differ, but we can still consolidate the return excess logic to avoid duplication and make the flow cleaner.

Debugger022 commented on contracts/SwapRouter/SwapRouter.sol:
```diff
@@ -0,0 +1,493 @@
+    ) internal returns (uint256 amountOut) {
+        // If tokens are the same, no swap needed
+        if (tokenIn == tokenOut) {
+            if (minAmountOut > 0 && amountIn < minAmountOut) {
```
@GitGuru7 You're right that centralized validation is generally better, but in this case `_performSwap` is shared by 5 external functions, so putting the `minAmountOut > 0` check in each external function would actually create more duplication than keeping it in the shared internal function.

Also, `swapAndRepayFull` intentionally passes minAmountOut = 0 since it validates against debt amount instead, so the current approach in _performSwap with `if (minAmountOut > 0 && amountIn < minAmountOut)` properly handles this edge case without forcing each caller to know about this implementation detail.

Debugger022 committed:
fix: fix comments

fred-venus commented on contracts/SwapRouter/SwapRouter.sol:
```diff
@@ -0,0 +1,486 @@
+        uint256 amountIn,
+        uint256 minAmountOut,
+        bytes calldata swapCallData
+    ) external payable nonReentrant {
```
can remove payable here

Debugger022 committed:
fix: major fixes

Debugger022 commented on contracts/SwapRouter/SwapRouter.sol:
```diff
@@ -0,0 +1,486 @@
+    /**
+     * @notice Accepts native tokens sent to this contract
+     */
+    receive() external payable {}
```
[f2459c8](https://github.com/VenusProtocol/venus-periphery/pull/16/commits/f2459c8742ad2f4a2075b0485ebc247ea1d284c7)

Debugger022 commented on contracts/SwapRouter/SwapRouter.sol:
```diff
@@ -0,0 +1,493 @@
+     * @param amount The amount to supply
+     * @return amountSupplied The actual amount supplied
+     */
+    function _supply(address vToken, address token, uint256 amount) internal returns (uint256 amountSupplied) {
```
[f2459c8](https://github.com/VenusProtocol/venus-periphery/pull/16/commits/f2459c8742ad2f4a2075b0485ebc247ea1d284c7)

Exef commented on contracts/SwapRouter/SwapRouter.sol:
```diff
@@ -0,0 +1,498 @@
+        uint256 minAmountOut,
+        bytes calldata swapCallData
+    ) external payable nonReentrant {
+        if (msg.value == 0) revert ZeroAmount();
```
Since this function uses `mintBehalf` it might be a good idea to check if `msg.sender` delegated this contract.

Debugger022 replied:
`mintBehalf` function doesn't require delegation logic.

Exef commented on contracts/SwapRouter/SwapRouter.sol:
```diff
@@ -0,0 +1,498 @@
+        uint256 minAmountOut,
+        bytes calldata swapCallData
+    ) external nonReentrant {
+        if (amountIn == 0) revert ZeroAmount();
```
Since this function uses mintBehalf it might be a good idea to check if msg.sender delegated this contract.

Exef commented on contracts/SwapRouter/SwapRouter.sol:
```diff
@@ -0,0 +1,498 @@
+        uint256 minAmountOut,
+        bytes calldata swapCallData
+    ) external payable nonReentrant {
+        if (msg.value == 0) revert ZeroAmount();
```
Since this function uses mintBehalf it might be a good idea to check if msg.sender delegated this contract.

Exef commented on contracts/SwapRouter/SwapRouter.sol:
```diff
@@ -0,0 +1,498 @@
+        uint256 maxAmountIn,
+        bytes calldata swapCallData
+    ) external payable nonReentrant {
+        if (maxAmountIn == 0) revert ZeroAmount();
```
Since this function uses mintBehalf it might be a good idea to check if msg.sender delegated this contract.

Debugger022 committed:
fix: fix unit and fork test

Debugger022 committed:
Merge branch 'feat/swapper' into feat/swapRouter

Debugger022 commented on contracts/SwapRouter/SwapRouter.sol:
```diff
@@ -0,0 +1,486 @@
+     * @dev For native vToken (vBNB), returns wrapped native token address
+     */
+    function _getUnderlyingToken(address vToken) internal view returns (address underlying) {
+        if (vToken == address(NATIVE_VTOKEN)) {
```
It will be WBNB. However, for VBNB, we are also returning WBNB.

Exef committed:
Merge branch 'develop' into feat/swapRouter-develop-merge

Debugger022 committed:
test: update unit test for swapRouter

Debugger022 committed:
Merge pull request #31 from VenusProtocol/feat/swapRouter-develop-merge

Swap Router - merge develop to update SwapHelper

Debugger022 committed:
fix: tests

Debugger022 committed:
test: update swapRouter fork tests

Debugger022 committed:
feat: add swapRouter bscmainnet deployment

Debugger022 committed:
feat: updating deployment files

Debugger022 committed:
feat: add swapNativeAndRepayFull function

Debugger022 committed:
feat: L-01

Debugger022 committed:
feat: L-05

Debugger022 committed:
refactor: consolidate slippage check in repayFull

Debugger022 committed:
feat: I-01

Debugger022 committed:
feat: add SwapRouter audits

Debugger022 committed:
Merge pull request #45 from VenusProtocol/fix/VPD-458

[VPD-458]: Cyfrin Audit Fixes for SwapRouter

Debugger022 committed:
Merge pull request #43 from VenusProtocol/fix/vpd-440

[VPD-440]: Pashov Audit Fixes for SwapRouter

Debugger022 committed:
feat: add deployments

Debugger022 committed:
feat: updating deployment files

Debugger022 committed:
Merge branch 'develop' into feat/swapRouter

Debugger022 committed:
feat: add barrel Interfaces.sol to re-export split interface files

Prevents duplicate artifact errors (HH701) by making Interfaces.sol
a re-export file pointing to contracts/Interfaces/*.sol, preserving
the import path used by SwapRouter without requiring redeployment.

Debugger022 committed:
feat: updating deployment files

Debugger022 committed:
chore: clean up .gitignore duplicates and organize sections

fred-venus commented:
https://venus.io/#/governance/proposal/593?chainId=56
